### DISCUSSION TITLE
How to create image radius like css border-radius

### DISCUSSION THREAD

**Discussion by demisweety** [libvips/libvips#2397] [Help wanted]
![image](https://user-images.githubusercontent.com/41557487/129441951-468a2995-f767-40d0-a1e1-328d73d9e37e.png)
Can give me a C or Python example!

**Comment by jcupitt** [comment]
Hello, do you need to create exactly that image?

It'd probably be simplest with SVG, eg.:

```python
#!/usr/bin/python3

import pyvips

x = pyvips.Image.svgload_buffer(b"""
<svg viewBox="0 0 200 200">
  <circle r="100" cx="100" cy="100" fill="#900"/>
</svg>
""")

x.write_to_file("x.png")
```

Though drawing a rect instead, obviously.

**Reply by demisweety** [reply]
am using this method, but I need to keep the PNG format transparent to have an effect. After the image is converted to PNG, the size is relatively large and the efficiency is not high, can you give a better suggestion.

**Comment by jcupitt** [comment]
I suppose you could try lossless webp, it's a little better than png.

**Reply by demisweety** [reply]
Now there is a phenomenon, I first rotate and then write buffer, 1m picture tps 1000 in the case of high concurrency pressure test, soon the host memory is used up, and then it crashes

**Reply by jcupitt** [reply]
Post some sample code in a new discussion that produces high memory use and I'll have a look.

**Reply by demisweety** [reply]
tks for your help! I resolved this problem! I forgot to release resource!

**Reply by demisweety** [reply]
When I used svg to implement this function, this crash occurred under high concurrency。
![image](https://user-images.githubusercontent.com/41557487/129481415-7806678a-46d6-4f47-a0a7-c26269cd5e76.png)

jni core：
JNIEXPORT void JNICALL
Java_com_criteo_vips_VipsImage_radiusNative(JNIEnv *env, jobject obj, jint radius) {
    VipsImage *image = (VipsImage *) (*env)->GetLongField(env, obj, handle_fid);
    char *svg;
    VipsImage *alpha;
    VipsImage *x;

    svg = g_strdup_printf("<svg viewBox=\"0 0 %1$d %2$d\">"
                          "<rect rx=\"%3$d\" ry=\"%3$d\" x=\"0\" y=\"0\" width=\"%1$d\" height=\"%2$d\" fill=\"#fff\"/>"
                          "</svg>", image->Xsize, image->Ysize, radius);
    if (!(alpha = vips_image_new_from_buffer(svg, strlen(svg), "", NULL))) {
        g_object_unref(image);
        g_free(svg);
        throwVipsException(env, "Unable to vips_image_new_from_buffer blur...");
    }
    g_free(svg);

    /* Get the alpha from the SVG image.
     */
    if (vips_extract_band(alpha, &x, 3, NULL)) {
        g_object_unref(image);
        g_object_unref(alpha);
        throwVipsException(env, "Unable to vips_extract_band blur...");
    }
    g_object_unref(alpha);
    alpha = x;

    /* Join to the main image.
         */
    if (vips_bandjoin2(image, alpha, &x, NULL)) {
        g_object_unref(image);
        g_object_unref(alpha);
        throwVipsException(env, "Unable to vips_bandjoin2 blur...");
    }
    g_object_unref(alpha);
    g_object_unref(image);
    image = x;

    (*env)->SetLongField(env, obj, handle_fid, (jlong) image);
    g_object_unref(image);
}

**Reply by demisweety** [reply]
![image](https://user-images.githubusercontent.com/41557487/129481579-c0904400-6330-4aa4-99fb-8271a3e5b744.png)

**Reply by jcupitt** [reply]
I would guess your ref counts are messed up. You're probably unreffing `image` too early.
